### PR TITLE
Migrate top insert toolbar to notebook view zone

### PR DESCRIPTION
**Pull request by rebornix**

### CODE REVIEW

File: src/vs/workbench/contrib/notebook/browser/controller/insertCellActions.ts [hunk 1]
```diff
@@ -227,6 +227,50 @@ registerAction2(class InsertMarkdownCellAtTopAction extends NotebookAction {
 	}
 });
 
+registerAction2(class extends NotebookAction {
+	constructor() {
+		super(
+			{
+				id: 'notebook.actions.insertWhitespace',
+				title: localize('notebook.actions.insertWhitespace', "Add Whitespace"),
```
Comment (Tyriar) [inline src/vs/workbench/contrib/notebook/browser/controller/insertCellActions.ts:235]: Insert vs Add intentional?

File: src/vs/workbench/contrib/notebook/browser/controller/insertCellActions.ts [hunk 2]
```diff
@@ -227,6 +227,50 @@ registerAction2(class InsertMarkdownCellAtTopAction extends NotebookAction {
...
+			this.runWithContext(accessor, context);
+		}
+	}
+
+	async runWithContext(accessor: ServicesAccessor, context: INotebookActionContext): Promise<void> {
+		if (context.cell) {
+			const index = context.notebookEditor.getCellIndex(context.cell);
+			if (index >= 0) {
+				const span = document.createElement('span');
+				span.innerText = 'Hello whitespace';
```
Comment (Tyriar) [inline src/vs/workbench/contrib/notebook/browser/controller/insertCellActions.ts:257]: test string?

File: src/vs/workbench/contrib/notebook/browser/view/notebookCellListView.ts [hunk 3]
```diff
@@ -147,6 +147,31 @@ export class NotebookCellsLayout implements IRangeMap {
 		}
 	}
 
+	/**
+	 * find position of whitespace
+	 * @param id
+	 * @returns
```
Comment (Tyriar) [inline src/vs/workbench/contrib/notebook/browser/view/notebookCellListView.ts:153]: Partial docs?

File: src/vs/workbench/contrib/notebook/browser/viewParts/notebookViewZones.ts [hunk 4]
```diff
@@ -29,25 +31,39 @@ export class NotebookViewZones extends Disposable {
...
 				return this._addZone(zone);
 			},
 			removeZone: (id: string): void => {
+				zonesHaveChanged = true;
+				// TODO: validate if zones have changed layout
 				this._removeZone(id);
 			},
 			layoutZone: (id: string): void => {
+				zonesHaveChanged = true;
+				// TODO: validate if zones have changed layout
```
Comment (Tyriar) [inline src/vs/workbench/contrib/notebook/browser/viewParts/notebookViewZones.ts:54]: Intentionally not included in PR?

**Review by Tyriar — DISMISSED** [review]

File: src/vs/workbench/contrib/notebook/browser/controller/insertCellActions.ts (hunk 1 above)
Comment (rebornix) [reply src/vs/workbench/contrib/notebook/browser/controller/insertCellActions.ts:235]: Good catch, this should be removed.

File: src/vs/workbench/contrib/notebook/browser/viewParts/notebookViewZones.ts (hunk 4 above)
Comment (rebornix) [reply src/vs/workbench/contrib/notebook/browser/viewParts/notebookViewZones.ts:54]: Right, this one needs more thinking, I plan to tackle this later.

**Review by rzhao271 — APPROVED** [review]

**Pull request merged**